### PR TITLE
MB-66210: LineString and MultiLineString changes

### PR DESCRIPTION
**Pull request by Likith101**
- Changed polylines intersects polygons function to first make a shape index of the polygons and then perform a series of contains point queries to identify if any of the vertices fall inside the polygon. If this doesn't yield a result, it means that there is an edge crossing with no vertex overlap, so we perform all edge intersect queries
 - Removed unnecessary loop within edge crossing logic
 - Changed polyline intersects envelope by checking each vertex's bounds and then doing an edge crossing check
 - Removed unecessary loop within this edge crossing logic as well
 - Fixed polyline intersects circle logic. Previously would only check a single edge of the polyline before exiting the function
 - Added contains logic for points and multipoints as they are very similar to intersects.
 - Added appropriate test cases covering edge cases
 
Note
    LineString contains LineString is theoretically possible, but there is no native support for this and trying to come up with an implementation myself proved futile. Perhaps we can consider this as a TODO for the future.

### CODE REVIEW

File: geojson/geojson_shapes_impl.go [hunk 1]
```diff
@@ -1156,6 +1157,27 @@ func checkLineStringsIntersectsShape(pls []*s2.Polyline, shapeIn,
...
+		// check the containment for every point in the collection.
+		for _, point := range p2.s2points {
+			if !polylineIntersectsPoint(pls, point) {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
```
Comment (CascadingRadium) [inline geojson/geojson_shapes_impl.go:1170]: can't a linestring `contain` another `linestring` as well?
For example ->
``` 
if i have line [[0,0],[2,0]] <X,Y coords>
it can contain line [[0,0],[1,0]]
```

Comment (Likith101) [reply geojson/geojson_shapes_impl.go:1170]: The library has no existing support or math for this particular operation. While I do agree that it is possible, coming up with it myself proved futile.

Comment (CascadingRadium) [reply geojson/geojson_shapes_impl.go:1170]: it is possible no? if all the vertex in other polyline is present in the original polyline, then it can be done, reusing the Project method. (as Polyline is array of point)

```
	if ols, ok := other.(*LineString); ok {
		opl := ols.pl
		for _, pl := range pls {
			// check if pl is containing opl
			// so all vertex/point in opl should be in pl
			contains := true
			for _, oplVertex := range opl.Vertices() {
				if !polylineIntersectsPoint([]*s2.Polyline{pl}, &oplVertex) {
					contains = false
					break
				}
			}
			if contains {
				return true, nil
			}
		}
	}


func (p *Polyline) Vertices() []Point {
	if p == nil {
		return nil
	}
	return *p
}
```

**Review by CascadingRadium — DISMISSED** [review]
lgtm, can add the linestring contains linestring/multilinestring as optional

File: geojson/geojson_shapes_impl.go (hunk 1 above)
Comment (Likith101) [reply geojson/geojson_shapes_impl.go:1170]: Just vertex overlaps does not mean contains. Take a 2d example of (2,0), (0,0), (0,2) and (1,0), (0,1). Both vertices of the second polyline are on the first, but the first one does not contain the second.

**Review by CascadingRadium — APPROVED** [review]

**Review by abhinavdangeti — APPROVED** [review]
